### ISSUE TITLE
False positive in `unnecessary_to_owned` when a mutable borrow would cause a compiler error

### ISSUE THREAD

**Issue by shssoichiro** [rust-lang/rust-clippy#14918]
### Summary

Minimal reproduction:

```rust
pub fn one_borrow(data: &mut [u8]) {
    multiple_borrows(
        // warning: unnecessary use of `to_vec`
        // but if we remove `to_vec()`, we get
        // error[E0502]: cannot borrow `*data` as mutable because it is also borrowed as immutable
        &data[0..8].to_vec(),
        &mut data[8..16]
    );
}

fn multiple_borrows(src:&[u8], dest: &mut [u8]){
    // Doesn't really matter what's in this function
    if src.len() == dest.len() {
        dest.copy_from_slice(src);
    }
}
```

(I know in this contrived use case you can use `split_at_mut`, but this can occur in cases where that isn't possible)

### Lint Name

unnecessary_to_owned

### Reproducer

I tried this code:

```rust
pub fn one_borrow(data: &mut [u8]) {
    multiple_borrows(
        // warning: unnecessary use of `to_vec`
        // but if we remove `to_vec()`, we get
        // error[E0502]: cannot borrow `*data` as mutable because it is also borrowed as immutable
        &data[0..8].to_vec(),
        &mut data[8..16]
    );
}

fn multiple_borrows(src:&[u8], dest: &mut [u8]){
    // Doesn't really matter what's in this function
    if src.len() == dest.len() {
        dest.copy_from_slice(src);
    }
}
```

I saw this happen:

```
warning: unnecessary use of `to_vec`
```

I expected to see this happen:

No warning.

### Version

```text
occurs on both rustc 1.87.0 (17067e9ac 2025-05-09) and rustc 1.89.0-nightly (283db70ac 2025-05-25)

rustc 1.89.0-nightly (283db70ac 2025-05-25)
binary: rustc
commit-hash: 283db70ace62a0ae704a624e43b68c2ee44b87a6
commit-date: 2025-05-25
host: x86_64-unknown-linux-gnu
release: 1.89.0-nightly
LLVM version: 20.1.5
```

### Additional Labels

@rustbot label +I-suggestion-causes-error

**Comment by blyxyas** [comment]
Bisect result: This false positive has been here all the way from 1.59.0 (when this lint was added).

**Comment by youknowone** [comment]
Another similar false positive case:

```rust
fn main() {
    let slice = &[1, 2, 3];
    let _ = if true {
        // suggestion: slice.iter().copied()
        slice.to_vec().into_iter()
    } else {
        vec![1].into_iter()
    };
}
```

```
warning: unnecessary use of `to_vec`
 --> t.rs:5:9
  |
5 |         slice.to_vec().into_iter()
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^ help: use: `slice.iter().copied()`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_to_owned
  = note: `#[warn(clippy::unnecessary_to_owned)]` on by default

warning: 1 warning emitted
```

Applying the suggestion break builds.

**Comment by code-with-aneesh** [comment]
Can i work on this issue ??
It is assigned to someone but they have not posted an update for 2 months.

**Comment by ShoyuVanilla** [comment]
Oh, sorry. I forgot about this as I was working on some issues in rust-analyzer. Please go ahead! IMO this should be merged with another similar existing lint that correctly handles borrowck though I cannot remember the precise name 😅 
@rustbot release-assignment

**Comment by code-with-aneesh** [comment]
Thanks !
Do you have any tips for me to tackle this ?
I am a little new to rust-clippy

**Comment by ShoyuVanilla** [comment]
Well, as I'm not a clippy team member I cannot give the really useful advices 😅 (And the above merging into existing lint, `redundant _clone` - I recalled its name - is just a personal opinion as well) I guess you could get some idea from the `redundant_clone` lints or share some common functionalities with it. And if you happen to get stuck, asking on rustlang zulip's clippy channel would be the best way to get some help

**Comment by code-with-aneesh** [comment]
Thank You Very Much 🙏🙏
